[CK-Tile] Enable vectorized reads on all layouts & improve perf.

aosewski opened the pull request:
## Proposed changes

This PR:

- enables vectorized reads on all layouts for input tensors {RR,RC,CC,CR} 
- Improve ck-tile universal gemm compute bound pipeline perf by from ~4% to 26%.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

aosewski committed:
Refactor universal gemm policy.

aosewski committed:
Adapt example to refactor changes.

aosewski committed:
Introduce static encoding pattern

aosewski committed:
Adding shuffled encoding patterns.

aosewski committed:
Fix err in reverse tuple.

aosewski committed:
Add transpose_tile2d

aosewski committed:
Small refactoring + doc

aosewski committed:
Enable reading on contiguous dimension in all layouts.

aosewski committed:
Transpose A/B register tile if needed for comp v3 pipeline.

aosewski committed:
Take contiguous dim size when calculating dram vector load size.

aosewski committed:
A/B smem pack size taken from WarpGemm attributes

aosewski committed:
Update B LDS layout and setup tile distribution pattern at class level.

aosewski committed:
Fix static assert.

aosewski committed:
Merge remote-tracking branch 'origin/develop' into aosewski/ck_tile_gemm_policy

aosewski committed:
Fix errors in examples.

aosewski committed:
Formatting & fix IsTranspose

aosewski committed:
Fix VectorSize & refactor.

aosewski committed:
Add error loging messages.

aosewski committed:
Fix VecLoadSize and TranspseC for mem pipeline.

aosewski committed:
Update unit-tests & disable mem pipeline.

aosewski committed:
Merge branch 'develop' into aosewski/ck_tile_gemm_policy

aosewski committed:
Clang format

jakpiase commented on example/ck_tile/03_gemm/gemm_basic.cpp:
```diff
@@ -103,4 +101,26 @@ float gemm_calc(const ck_tile::GemmHostArgs& args, const ck_tile::stream_config&
+    std::string a_layout = arg_parser.get_str("a_layout");
+    std::string b_layout = arg_parser.get_str("b_layout");
+
+    if(a_layout == "R" && b_layout == "C")
```
Why do we suuport only for RC layout?

aosewski replied:
With current GemmKernel design with default pipelines it works only for this layout. Default pipelines force K-major tiles and this way they can't do vector reads on non-K major tensors. Current GemmKernel leverages all layouts and fosters vectorized reading on any contiguous dimension no matter it's K or not.

jakpiase commented on include/ck_tile/core/algorithm/static_encoding_pattern.hpp:
```diff
@@ -0,0 +1,210 @@
+    block_raked,
+};
+
+struct TileDistributionEcodingPattern
```
A typo here - should be TileDistributionEncodingPattern

jakpiase commented on include/ck_tile/core/tensor/tile_window.hpp:
```diff
@@ -18,8 +18,17 @@
+/**
+ * @brief This class provides tile (windowed) view and access to the device memory.
+ *
+ * @note this tile window do not support single issue you need to use tile_window_linear
```
```suggestion
 * @note This tile window does not support single issue you need to use tile_window_linear
```

jakpiase commented on include/ck_tile/core/tensor/transpose_tile.hpp:
```diff
@@ -0,0 +1,201 @@
+        return y_dim_out_to_in_;
+    }();
+
+    //
```
A comment is probably missing here

jakpiase commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -250,6 +275,14 @@ struct GemmKernel
                                                    const GemmKernelArgs& kargs,
                                                    const SplitKBatchOffset& splitk_batch_offset)
     {
+        // const auto idxs = TilePartitioner{}();
+        // const auto i_m  = idxs.at(number<0>{});
+        // const auto i_n  = idxs.at(number<1>{});
+        // // options
+        // const ADataType* a_start = static_cast<const ADataType*>(kargs.a_ptr);
+        // const BDataType* b_start = static_cast<const BDataType*>(kargs.b_ptr);
+        // // Convert pointers to tensor views
+        // auto a_tensor_view = [&]() {
```
Probably leftovers

aosewski replied:
Indeed!

jakpiase commented on include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp:
```diff
@@ -99,54 +233,193 @@ struct UniversalGemmPipelineAgBgCrPolicy
+            return b_lds_block_desc;
+        }
+#else
+        else // B is Row Major
```
Why is all the code below commented?

aosewski replied:
Code left for future optimization. Should be changed in next PRs. Currently works better with this first part.

jakpiase reviewed (commented):
Great work , I have left some minor comments

bartekxk commented on include/ck_tile/core/tensor/transpose_tile.hpp:
```diff
@@ -0,0 +1,201 @@
+    constexpr index_t num_vec_out = vec_length_in;
+
+    using InVec  = array<DataType, vec_length_in>;
+    using OutVec = array<DataType, vec_length_out>;
```
Should we add static assert that InputDataType and OutputDataType are same?

bartekxk commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v3.hpp:
```diff
@@ -318,11 +359,31 @@ struct GemmPipelineAgBgCrCompV3 : public BaseGemmPipelineAgBgCrCompV3<Problem>
-                    Base::GlobalPrefetch(b_block_tile, b_copy_dram_window);
+                    if constexpr(is_a_col_major)
+                    {
+                        auto a_shuffle_tmp = make_static_distributed_tensor<ADataType>(
```
Maybe it is worth to create static_distributed_tensor for shuffle outside the tuple and reuse if needed. You can use conditional_t and set it to empty tuple if not needed. (The same for B)

aosewski replied:
Would need to check which option uses more registers.

aosewski replied:
Checked: both versions use exactly the same amount of registers and yields same perf. So I decided to leave current impl.

bartekxk commented on test/ck_tile/gemm/test_gemm_pipeline.cpp:
```diff
@@ -14,26 +14,28 @@ using Row       = ck_tile::tensor_layout::gemm::RowMajor;
 using KernelTypes = ::testing::Types<
     //         ALayout, BLayout, CLayout, ADataType, BDataType, AccDataType, CDataType, GemmPipelineScheduler, PipelineType
-    std::tuple<    Row,     Row,     Row,       F16,       F16,         F32,       F16,             Intrawave,         Mem>,
+    // std::tuple<    Row,     Row,     Row,       F16,       F16,         F32,       F16,             Intrawave,         Mem>,
```
WHy commented?

aosewski replied:
Cause Mem pipline is not yet updated to have vectorized reads on non-k-major tensors. Would be added in next PR

bartekxk reviewed (commented):
LGTM

aosewski committed:
Update include/ck_tile/core/tensor/tile_window.hpp

Co-authored-by: jakpiase <jakub.piasecki@amd.com>

aosewski committed:
Merge branch 'develop' into aosewski/ck_tile_gemm_policy

aosewski committed:
Fix compilation and reviewers comments.

jakpiase reviewed (dismissed):
LGTM, well done

aosewski committed:
Refactor unit-test. Fallback to non-universal gemm.

Need to use GemmPipelineAGmemBGmemCRegV1 for now,
since GemmKernel is now supporting also non-K major vector reads.

aosewski committed:
Merge branch 'develop' into aosewski/ck_tile_gemm_policy
